lib/AST/StmtManual.cpp

```diff
@@ -19,6 +19,95 @@ namespace pasta {
 
 #ifndef PASTA_IN_BOOTSTRAP
 
+// Follow a macro token's parent chain to the end, and returns the final parent.
+// If the macro token has no parent, this returns std::nullopt;
+std::optional<pasta::Macro>
+GetRootSubstitution(const pasta::MacroToken &macro_token) noexcept {
+  auto root = std::optional(macro_token.Parent());
+  while (root && root->Parent()) {
+    root = root->Parent();
+  }
+  return root;
+}
+
+// Returns the highest substitution that covers this Stmt, if any.
+std::optional<pasta::Macro> Stmt::GetCoveringSubstitution() const noexcept {
+  // If the first token in this Stmt did not come from a macro substitution,
+  // then this Stmt is not covered by a substitution
+  const auto begin_macro_token = BeginToken().MacroLocation();
+  if (!begin_macro_token) {
+    return std::nullopt;
+  }
+
+  // If the last token in this Stmt did not come from a macro substitution,
+  // then this Stmt is not covered by a substitution
+  const auto end_macro_token = BeginToken().MacroLocation();
```

Comment:
`EndToken`? I think you probably want `.Tokens()` then get front/back, or `[0]` and `[size - 1u]`.

You might also want `AST::Containing(*this).Tokens()`, and then use `.Index()` of the stmt front/back tokens to check the prev/next tokens in the AST (skipping over non-parsed roles).

Reply:
Is `AST::Containing()` already defined? I'm having trouble finding it.

Reply:
Ah! Maybe add an `AST::From` for `Stmt`, like this one: https://github.com/trailofbits/pasta/blob/master/include/pasta/AST/AST.h#LL49C1-L49C1